sycl/test-e2e/Matrix/joint_matrix_prefetch_impl.hpp

```diff
@@ -0,0 +1,141 @@
+#define TM 8
+#define TK 16
+
+template <typename T1, typename T2, typename T, size_t M, size_t N, size_t K,
+          unsigned int vnniFactor>
+void joint_matrix_gemm_vnni(sub_group sg, size_t sg_startx, size_t sg_starty,
+                            T1 *A, T2 *B, T *C) {
+  auto pA = address_space_cast<sycl::access::address_space::global_space,
+                               access::decorated::no>(A);
+  auto pB = address_space_cast<sycl::access::address_space::global_space,
+                               access::decorated::no>(B);
+  auto pC = address_space_cast<sycl::access::address_space::global_space,
+                               access::decorated::no>(C);
+
+  joint_matrix<sub_group, T1, use::a, TM, TK, layout::row_major> sub_a;
+  joint_matrix<sub_group, T2, use::b, TK, TN, layout::ext_intel_packed> sub_b;
+  joint_matrix<sub_group, T, use::accumulator, TM, TN> sub_c;
+  joint_matrix_prefetch(sg, A + (sg_startx * TM) * K, K, layout::row_major, TM,
+                        TK, syclex::properties{syclex::prefetch_hint_L1});
+  joint_matrix_prefetch(sg, B + sg_starty / SG_SZ * TN * vnniFactor,
+                        N * vnniFactor, layout::ext_intel_packed,
+                        TK / vnniFactor, TN * vnniFactor,
+                        syclex::properties{syclex::prefetch_hint_L1});
+  joint_matrix_prefetch(sg, C + (sg_startx * TM) * N + sg_starty / SG_SZ * TN,
+                        N, layout::row_major, TM, TN,
+                        syclex::properties{syclex::prefetch_hint_L1});
+  joint_matrix_fill(sg, sub_c, 1);
+  for (int k = 0; k < K; k += TK) {
+    joint_matrix_load(sg, sub_a, pA + (sg_startx * TM) * K + k, K);
+    joint_matrix_load(sg, sub_b,
+                      pB + k * N + sg_starty / SG_SZ * TN * vnniFactor,
+                      N * vnniFactor);
+    joint_matrix_mad(sg, sub_c, sub_a, sub_b, sub_c);
+  }
+  joint_matrix_store(sg, sub_c,
+                     pC + (sg_startx * TM) * N + sg_starty / SG_SZ * TN, N,
+                     layout::row_major);
+}
+
+template <typename T1, typename T2, typename T, size_t M, size_t N, size_t K>
+void joint_matrix_gemm(sub_group sg, size_t sg_startx, size_t sg_starty, T1 *A,
```

Comment:
I think joint_matrix_gemm and joint_matrix_gemm_vnni can be unified, by adding a few more parameters.
We can use joint_matrix_gemm_vnni version and pass vnniFactor=1 for non-vnni case, also new parameter is needed to pass layout::row_major or layout::ext_intel_packed for matrix B.
Finally, need to select correct syclex::prefetch_hint. I don't know, what is better: pass it as a parameter as well or add if/else condition inside function.